src/int.cr

```diff
@@ -2776,3 +2776,45 @@ struct UInt128
     self
   end
 end
+
+# Returns a number for given digits and base.
+# The digits are expected as an Enumerable with the least significant digit as the first element.
+#
+# Base must not be less than 2.
+#
+# All digits must be within 0...base.
+#
+# ```
+# Int32.from_digits([5, 4, 3, 2, 1])           # => 12345
+# Int32.from_digits([4, 6, 6, 0, 5], base: 7)  # => 12345
+# Int32.from_digits([45, 23, 1], base: 100)    # => 12345
+#
+# Int32.from_digits([1], base: -2) # => ArgumentError
+# Int32.from_digits([-1]) # => ArgumentError
+# Int32.from_digits([3], base: 2) # => ArgumentError
+# ```
+{% for type in %w(Int8 Int16 Int32 Int64 Int128 UInt8 UInt16 UInt32 UInt64 UInt128) %}
+  def {{type.id}}.from_digits(digits : Enumerable(Int), base : Int = 10) : self
+    if base < 2
+      raise ArgumentError.new("Invalid base #{base}")
+    end
+
+    num : {{type.id}} = 0
+    multiplier = 1
+
+    digits.each do |digit|
+      if digit < 0
+        raise ArgumentError.new("Invalid digit #{digit}")
+      end
+
+      if digit >= base
+        raise ArgumentError.new("Invalid digit #{digit} for base #{base}")
+      end
+
+      num += digit * multiplier
+      multiplier *= base
```

Comment:
Multiplication should not be done for the last digit, otherwise this would overflow for numbers near `type`'s maximum, e.g. `Int32.from_digits([0xFF, 0xFF, 0xFF, 0x7F], base: 256)` would set `multiplier` to `0x100000000` for the last loop. Ditto if `multiplier` is no longer hardcoded to `Int32`